Description: Protect NTL coefficient calls from PARI's coeff macro
 Cython can include pari.h before ntlwrap_impl.h and cypari.h afterwards.
 In that order PARI's three-argument coeff macro intercepts NTL's
 two-argument function calls. Parenthesizing the function name prevents
 macro expansion without relying on the generated header order.
Origin: backport, https://github.com/sagemath/sage/commit/cd2cb16d82d549e21bdbe1361aa3c73f9646f43d
Author: Dima Pasechnik <dima@pasechnik.info>
Forwarded: not-needed
Last-Update: 2026-10-11

--- a/src/sage/libs/ntl/ntlwrap_impl.h
+++ b/src/sage/libs/ntl/ntlwrap_impl.h
@@ -10,6 +10,9 @@
 
 #include <NTL/ZZ_limbs.h>
 
+// PARI also defines coeff(a, i, j) as a macro. Parenthesize NTL's coeff
+// below so both libraries can be included in the same Cython module.
+
 //////// ZZ //////////
 
 /* Return value is only valid if the result should fit into an int.
@@ -155,15 +158,15 @@
    AUTHOR: David Harvey (2006-06-08) */
 static CYTHON_INLINE int ZZX_getitem_as_int(struct ZZX* x, long i)
 {
-    return ZZ_to_int(&coeff(*x, i));
+    return ZZ_to_int(&(coeff)(*x, i));
 }
 
 /* Copies ith coefficient of x to output.
    Assumes output has been mpz_init'd.
    AUTHOR: David Harvey (2007-02) */
 static CYTHON_INLINE void ZZX_getitem_as_mpz(mpz_t output, struct ZZX* x, long i)
 {
-    const ZZ& z = coeff(*x, i);
+    const ZZ& z = (coeff)(*x, i);
     ZZ_to_mpz(output, &z);
 }
 
@@ -621,7 +624,7 @@
 {
     // Gets the p-adic valuation of the ith coefficient of f.
     ZZ *u = new ZZ();
-    long ans = ZZ_remove(*u, rep(coeff(f, i)), p);
+    long ans = ZZ_remove(*u, rep((coeff)(f, i)), p);
     delete u;
     return ans;
 }
